E2E: add WordPress.com Pro plan specs.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR introduces a new spec that covers the WordPress.com Pro plan feature in the context of the onboarding flow.

Key changes:
- new spec file which creates a new WordPress.com Pro plan site with an existing user.
- change signature of SignupPickPlanPage.selectPlan to return details of the new site it creates.
- addition of new calls in RestAPIClient to delete a site.

Other housekeeping changes:
- renaming of methods in PlansPage to be less verbose.
- changes to CartCheckoutPage in order to select saved payment method.

#### Testing instructions

- trigger the Pre-Release Tests against this branch.
  - ensure that no failures are reported.
- in TeamCity, navigate to the **Tests** tab and ensure the test suite and/or steps from the new spec has been run.

Related to https://github.com/Automattic/wp-calypso/issues/62492

worldomonation committed:
Laying some groundwork first:
- shorten the method names for PlansPage and make corresponding changes to the
specs.
- change the SignupPickPlansPage to return the details of site that is created.

worldomonation committed:
New spec for WordPress.com Pro plans:
- implement API call to delete site.
- changes to CartCheckoutPage to support selection of existing payment cards.
- populate new spec to test the Pro plan.

Additional check on the RestAPIClient.

worldomonation committed:
Correct existing test for Legacy Plan renew workflow.

worldomonation commented on packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:
```diff
@@ -226,12 +229,34 @@ export class CartCheckoutPage {
 	 * @param {PaymentDetails} paymentDetails Object implementing the PaymentDetails interface.
 	 */
 	async enterPaymentDetails( paymentDetails: PaymentDetails ): Promise< void > {
+		// Click on the Credit or debit card input in order
+		// to expand the fields.
+		const cardInputLocator = this.page.locator( `span:has-text("Credit or debit card")` );
+		await cardInputLocator.click();
+
+		// Begin filling in the card details from
+		// top to bottom.
```
This change is only tangentially related; it is more of a QoL update to ensure the method won't flake out if an existing payment method is already selected (eg. if user has a saved card).

worldomonation commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -39,7 +39,8 @@ const selectors = {
 
 	// My Plans view
-	myPlanTitle: ( planName: LegacyPlans ) => `.my-plan-card__title:has-text("${ planName }")`,
+	myPlanTitle: ( planName: LegacyPlans | Plans ) =>
+		`.my-plan-card__title:has-text("${ planName }")`,
```
The Plan card in Upgrades > Plans > My Plan can show both names of Legacy and current plans.

worldomonation commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -104,6 +98,37 @@ export class PlansPage {
 		await showButtonLocator.waitFor();
 	}
 
+	/**
+	 * Validates that the provided plan name is the title of the active plan in the My Plan tab of the Plans page. Throws if it isn't.
+	 *
+	 * @param {LegacyPlans} expectedPlan Name of the expected plan.
+	 * @throws If the expected plan title is not found in the timeout period.
+	 */
+	async validateActivePlan( expectedPlan: LegacyPlans | Plans ): Promise< void > {
+		const expectedPlanLocator = this.page.locator( selectors.myPlanTitle( expectedPlan ) );
+		await expectedPlanLocator.waitFor();
+	}
+
+	/**
+	 * Validates that the provided tab name is the the currently active tab in the wrapper Plans page. Throws if it isn't.
+	 *
+	 * @param {PlansPageTab} expectedTab Name of the expected tab.
+	 * @throws If the expected tab name is not the active tab.
+	 */
+	async validateActiveTab( expectedTab: PlansPageTab ): Promise< void > {
+		// For mobile sized viewport, the currently selected tab name
+		// is hidden behind a pseudo-dropdown.
+		// Therefore the valicdation will look for hidden element.
+		const currentSelectedLocator = this.page.locator(
+			selectors.activeNavigationTab( expectedTab )
+		);
+		if ( envVariables.VIEWPORT_NAME === 'mobile' ) {
+			await currentSelectedLocator.waitFor( { state: 'hidden' } );
+		} else {
+			await currentSelectedLocator.waitFor();
+		}
+	}
+
```
This is merely a renaming and repositioning of these two methods. Nothing has changed syntactically.

worldomonation commented on packages/calypso-e2e/src/lib/pages/signup-pick-plan-page.ts:
```diff
@@ -25,7 +32,21 @@ export class SignupPickPlanPage {
-	async selectPlan( name: Plans ): Promise< void > {
-		await this.plansPage.selectPlan( name );
+	async selectPlan( name: Plans ): Promise< NewSiteDetails > {
+		const [ response ] = await Promise.all( [
+			this.page.waitForResponse( /.*sites\/new\?.*/ ),
+			this.plansPage.selectPlan( name ),
```
Similar to what was done in [UserSignupPage](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/user-signup-page.ts#L62-L83), the idea here is to intercept the response from the `/sites/new` endpoint then obtain key site information from the response.

The response contains three things we are interested in:
- id;
- site url;
- site name;

These will be key in the cleanup parts of the spec.

Note, a request to `/sites/new/` is triggered regardless of which plan the user selects in this screen:
![image](https://user-images.githubusercontent.com/6549265/170102810-256aa054-3396-4850-851a-6dda73f42b36.png)

worldomonation commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -180,6 +202,81 @@ export class RestAPIClient {
+	 */
+	async deleteSite(
+		expectedSiteDetails: SiteClosureDetails
+	): Promise< SiteDeletionResponse | null > {
```
I've chosen to define this function to return multiple types, but I am not sure how I feel about this.

On one hand, if the function needs to signify that no action was performed (for any reason) it is a trivial matter to return the `null` value. The calling function can check the return value with a simple `if ( response )` and that's it.

On the other hand, having multiple return types add to the mental workload. Additionally, it requires the calling method to first perform a check.

An alternative proposal is for the`deleteSite` method to always return a `SiteDeletionResponse`, even for invalid or no-ops. 

The interface for `SiteDeletionResponse` is defined as:

```
interface SiteDeletionResponse {
	ID: number;
	name: string;
	status: string;
}
```

By setting the value for `status` attribute to something custom like "noop" the calling method can simply check the status.

worldomonation commented:
@southp I'm not sure if you've reviewed end-to-end specs before, but as @dpasque is out for the next week and half I wanted to get the ball rolling on the reviews. I have a whole stack of PRs that depend on this PR.

worldomonation commented on packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:
```diff
@@ -226,12 +229,34 @@ export class CartCheckoutPage {
+	 *
+	 * @param {string} cardHolderName Name of the card holder associated with the payment method.
+	 */
+	async selectSavedCard( cardHolderName: string ): Promise< void > {
+		const selector = this.page.locator( selectors.existingCreditCard( cardHolderName ) ).first();
+
+		await selector.click();
+	}
+
```
This is a QoL change to avoid creating new payment methods each time.

worldomonation committed:
Revise the changeset some more:
- update error message.
- add new step to validate the media capacity.
- consolidate type interface

worldomonation commented:
The Pre-Release test failure is due to https://github.com/Automattic/wp-calypso/issues/64212 to which I have created the PR https://github.com/Automattic/wp-calypso/pull/64218 to address.

dpasque commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -55,6 +55,19 @@ export class MediaPage {
+	 *
+	 * @param {number} capacity Expected capacity in GB (gigabytess).
+	 */
+	async validateStorageCapacity( capacity: number ): Promise< boolean > {
```
Tiny style nit: traditionally, we've had `validate*` functions be `void` returning functions that throw an error if something isn't there.

Since this returns a boolean value, can we make the name indicate that? E.g. `hasStorageCapacity` or something like that?

worldomonation replied:
> we've had validate* functions be void returning functions that throw an error if something isn't there

Hmm, this hadn't occurred to me before!

You make a good point about the expected results from `validate*` type of methods - will be happy to make the change in a revision.

dpasque commented on packages/calypso-e2e/src/lib/pages/signup-pick-plan-page.ts:
```diff
@@ -24,8 +26,24 @@ export class SignupPickPlanPage {
+			throw new Error( 'Failed to create new site when selecting a plan at signup.' );
+		}
+
+		const responseBody: NewSiteResponse = JSON.parse( ( await response.body() ).toString() );
```
Optional FYI here -- it looks like the Playwright `Response` object has a `json` function that does all this for you! https://playwright.dev/docs/api/class-response#response-json

worldomonation replied:
Oh :lol:

dpasque reviewed (approved):
This spec is beautiful! 😍 Excellent work! Left two tiny optional nits. Otherwise, LGTM! 🚀

worldomonation committed:
Revise with comments from review:
- change name of method to `hasStorageCapacity`;
- use built-in JSON decoder method for Playwright response.

worldomonation committed:
Use expect

worldomonation committed:
Temporarily wrap a step in Signup:Free with try/catch.

worldomonation committed:
Use `waitFor` instead
